sdk/src/azure/core/az_mqtt5_rpc.c

```diff
@@ -30,21 +30,137 @@ AZ_NODISCARD static az_result _az_rpc_substitute_key_for_value(
     az_span value,
     az_span out_topic);
 
-AZ_NODISCARD bool az_span_topic_matches_sub(az_span sub, az_span topic)
+/**
+ * @brief Helper function to obtain the next level of a topic.
+ *
+ * @param topic Topic to get the next level of.
+ * @return An #az_span value containing the next level of the topic including the backslash.
+ */
+AZ_NODISCARD AZ_INLINE az_span _get_next_topic_level(az_span topic)
 {
-  bool ret;
-#if defined(TRANSPORT_MOSQUITTO)
-  if (MOSQ_ERR_SUCCESS
-      != mosquitto_topic_matches_sub((char*)az_span_ptr(sub), (char*)az_span_ptr(topic), &ret))
+  int32_t pos = az_span_find(topic, AZ_SPAN_FROM_STR("/"));
+  if (pos == -1)
   {
-    ret = false;
+    int32_t pos_null_char = az_span_find(topic, AZ_SPAN_FROM_STR("\0"));
+    if (pos_null_char == -1)
+    {
+      return topic;
+    }
+    return az_span_slice(topic, 0, pos_null_char);
   }
-#else // TRANSPORT_MOSQUITTO
-  (void)sub;
-  (void)topic;
-  ret = false;
-#endif
-  return ret;
+
+  return az_span_slice(topic, 0, pos + 1);
+}
+
+/**
+ * @brief Helper function to check if a topic has a backslash at the end of it.
+ *
+ * @param topic Topic to check.
+ * @return True if the topic has a backslash at the end of it, false otherwise.
+ */
+AZ_NODISCARD AZ_INLINE bool _has_backslash_in_topic(az_span topic)
+{
+  if (az_span_size(topic) == 0)
+  {
+    return false;
+  }
+  return az_span_find(topic, AZ_SPAN_FROM_STR("/")) == az_span_size(topic) - 1;
+}
+
+AZ_NODISCARD bool _az_span_topic_matches_filter(az_span topic_filter, az_span topic)
```

Comment:
This is too complicated: the RPC client and RPC server should comply with a fixed template instead of having a generic topic matcher.
The function shouldn't be more complicated than https://github.com/Azure/azure-sdk-for-c/blob/feature/v2/sdk/src/azure/iot/az_iot_hub_client_methods.c#L24C24-L24C70.